spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/FirestoreTemplate.java

```diff
@@ -75,32 +82,37 @@ public FirestoreTemplate(FirestoreStub firestore, String parent) {
 		this.databasePath = parent.substring(0, StringUtils.ordinalIndexOf(parent, "/", 4));
 	}
 
-	public <T> Mono<T> findById(Publisher idPublisher, Class<T> aClass) {
-		return findAllById(idPublisher, aClass).next();
+	public <T> Mono<Boolean> existsById(Publisher<String> idPublisher, Class<T> entityClass) {
+		return Flux.from(idPublisher).next()
+				.flatMap(id -> getDocument(id, entityClass, NAME_ONLY_MASK))
+				.map(d -> d == DocumentReadContext.EMPTY ? Boolean.FALSE : Boolean.TRUE)
+				//.onErrorReturn(throwable -> throwable.getMessage().startsWith("NOT_FOUND: Document"), Boolean.FALSE)
+				.onErrorMap(
+						throwable -> new FirestoreDataException("Unable to test for document existence", throwable));
+	}
+
+	public <T> Mono<T> findById(Publisher idPublisher, Class<T> entityClass) {
+		return findAllById(idPublisher, entityClass).next();
 	}
 
-	public <T> Flux<T> findAllById(Publisher idPublisher, Class<T> aClass) {
-		return ((Flux<String>) Flux.from(idPublisher)).flatMap(id -> {
-			FirestorePersistentEntity<?> persistentEntity = this.mappingContext.getPersistentEntity(aClass);
-			GetDocumentRequest request = GetDocumentRequest.newBuilder()
-					.setName(this.parent + "/" + persistentEntity.collectionName() + "/" + id).build();
-			return ObservableReactiveUtil.<Document>unaryCall(obs -> this.firestore.getDocument(request, obs));
-		}).onErrorMap(throwable -> new FirestoreDataException("Unable to find an entry by id", throwable))
-				.map(document -> PublicClassMapper.convertToCustomClass(document, aClass));
+	public <T> Flux<T> findAllById(Publisher idPublisher, Class<T> entityClass) {
+		return ((Flux<String>) Flux.from(idPublisher)).flatMap(id -> getDocument(id, entityClass, null))
+				.flatMap(t -> t == DocumentReadContext.EMPTY ? Mono.empty() : Mono.just(t))
```

Comment:
It seems like you must check `DocumentReadContext.EMPTY` after every getDocument call; it might be better to just have `getDocument` return `Mono<Document>` and if the document doesn't exist, `getDocument` would return Mono.empty. Then you can use `Mono.switchOnEmpty()` to check if a document exists or not.